### ISSUE TITLE
Bug in BayesianModel's fit method

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#352]
``` python3
from sklearn.datasets import load_iris

iris = load_iris()

data = pd.DataFrame(iris.data.astype('int'), columns=['x1', 'x2', 'x3', 'x4'])

data['y'] = iris.target

data = data.reindex(np.random.randint(0, 150, 150))

model = BayesianModel([('x1', 'y'), ('x2', 'y'), ('x3', 'y'), ('x4', 'y')])

model.fit(data)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-107-d9a8c16dbfe9> in <module>()
----> 1 model.fit(data)

/home/ankur/pgmpy/pgmpy/models/BayesianModel.py in fit(self, data)
    541             if not parents:
    542                 state_counts = data.ix[:, node].value_counts()
--> 543                 cpd = TabularCPD(node, node_card[node],
    544                                  state_counts.values[:, np.newaxis])
    545                 cpd.normalize()

/usr/local/lib/python3.4/dist-packages/pandas/core/indexing.py in __getitem__(self, key)
   1198     def __getitem__(self, key):
   1199         if type(key) is tuple:
-> 1200             return self._getitem_tuple(key)
   1201         else:
   1202             return self._getitem_axis(key, axis=0)

/usr/local/lib/python3.4/dist-packages/pandas/core/indexing.py in _getitem_tuple(self, tup)
   1416                 continue
   1417 
-> 1418             retval = getattr(retval, self.name)._getitem_axis(key, axis=axis)
   1419 
   1420             # if the dim was reduced, then pass a lower-dim the next time

/usr/local/lib/python3.4/dist-packages/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   1470                 self._is_valid_integer(key, axis)
   1471 
-> 1472             return self._get_loc(key, axis=axis)
   1473 
   1474     def _convert_to_indexer(self, obj, axis=0, is_setter=False):

/usr/local/lib/python3.4/dist-packages/pandas/core/indexing.py in _get_loc(self, key, axis)
     89 
     90     def _get_loc(self, key, axis=0):
---> 91         return self.obj._ixs(key, axis=axis)
     92 
     93     def _slice(self, obj, axis=0, typ=None):

/usr/local/lib/python3.4/dist-packages/pandas/core/frame.py in _ixs(self, i, axis)
   1726             """
   1727 
-> 1728             label = self.columns[i]
   1729             if isinstance(i, slice):
   1730                 # need to return view

/usr/local/lib/python3.4/dist-packages/pandas/core/index.py in __getitem__(self, key)
    911 
    912         if np.isscalar(key):
--> 913             return getitem(key)
    914 
    915         if isinstance(key, slice):

IndexError: index out of bounds
```

**Comment by palashahuja** [comment]
@ankurankan  Is this issue closed?

**Comment by ankurankan** [comment]
@palashahuja No, still throwing this error.

**Comment by liquidmetal** [comment]
Just to keep this issue updated, I'm seeing this error now. @ankurankan are you seeing the same error?

```
Traceback (most recent call last):
  File "./tester.py", line 49, in <module>
    model.fit(data)
  File "/work/personal/pgmpy/pgmpy/models/BayesianModel.py", line 557, in fit
    cpds_list = estimator.get_parameters()
  File "/work/personal/pgmpy/pgmpy/estimators/MLE.py", line 75, in get_parameters
    evidence_card=parent_card.astype('int'))
  File "/work/personal/pgmpy/pgmpy/factors/CPD.py", line 139, in __init__
    super(TabularCPD, self).__init__(variables, cardinality, values.flatten('C'))
  File "/work/personal/pgmpy/pgmpy/factors/Factor.py", line 104, in __init__
    raise ValueError("Values array must be of size: {size}".format(size=np.product(cardinality)))
ValueError: Values array must be of size: 540
```

**Comment by khalibartan** [comment]
I'm also getting the same error.

**Comment by ankurankan** [comment]
@liquidmetal @khalibartan I am also getting this error. Will look into this.

**Comment by liquidmetal** [comment]
I suspect this is because the data is originally continuous. The data is then forced to be discrete and not all possible values are observed in the discrete data (for example, only 1, 5, 7 might be available). This causes the "Values array must be of size: X".

It seems like this isn't supported just yet. A workaround could be to fill in the missing values with zero probabilities (or epsilon).

**Comment by chrisittner** [comment]
[The first error](https://github.com/pgmpy/pgmpy/issues/352#issue-64903149) is fixed since [this](https://github.com/pgmpy/pgmpy/commit/b455ca2fd04d28bf53f83c8928b622d93b7016f7) commit. [The 2nd error](https://github.com/pgmpy/pgmpy/issues/352#issuecomment-177468505), is fixed in #667.

**Comment by khalibartan** [comment]
@ankurankan Close this issue. Fixed.
